homeassistant/components/ipma/sensor.py

```diff
@@ -89,11 +112,29 @@ def __init__(
         IPMADevice.__init__(self, api, location)
         self.entity_description = description
         self._attr_unique_id = f"{self._location.station_latitude}, {self._location.station_longitude}, {self.entity_description.key}"
+        self._ipma_data: Any | None = None
 
     @Throttle(MIN_TIME_BETWEEN_UPDATES)
     async def async_update(self) -> None:
         """Update sensors."""
         async with asyncio.timeout(10):
-            self._attr_native_value = await self.entity_description.value_fn(
+            self._ipma_data = await self.entity_description.value_fn(
                 self._location, self._api
             )
+
+    @property
+    def native_value(self) -> str | None:
+        """Return the state of the sensor."""
+        if self.entity_description.value_extractor is not None:
```

Comment:
I was wonder if we could simplify this a bit and avoid having `value_extractor` and `extra_attr_fn` at all, by making `value_fn` return a tuple of `[state, attrs]`? The sensors that do not have attributes could return `[state, None]`

Then we could keep `async_update` simple:

```python
state, attrs = await self.entity_description.value_fn(
    self._location, self._api
)
self._attr_native_value = state
self._attr_extra_state_attributes = attrs
```

Reply:
Initially tried to avoid changing any existing code, but you are right! this is a much cleaner solution! Thanks